File: src/clj/herder/solver/event.clj
(ns herder.solver.types
  (:use [herder.solver.helpers])
  (:import
   [herder.solver.helpers EventDifficultyComparator]
   [herder.solver EventType]))

(gen-class
 :name ^{PlanningEntity {"difficultyComparatorClass" EventDifficultyComparator}} herder.solver.types.Event
 :prefix "event-"
 :init init
 :state state
 :constructors {[] []
                [java.util.UUID] []}
 :methods [[^{PlanningVariable {"valueRangeProviderRefs" ["slotRange"] "nullable" true}} getSlot [] Object]
           [setSlot [Object] void]
           [getPreferredSlots [] java.util.List]
           [setPreferredSlots [java.util.List] void]
           [getChainedEvent [] Object]
           [setChainedEvent [Object] void]
           [getEventDay [] int]
           [setEventDay [int] void]
           [getDependantEventCount [] int]
           [setDependantEventCount [int] void]
           [getEventType [] herder.solver.EventType]
           [setEventType [herder.solver.EventType] void]

           [getId [] java.util.UUID]
           [getExternalId [] java.util.UUID]
           [getPeople [] java.util.List]
           [setPeople [java.util.List] void]
           [getName [] String]
           [setName [String] void]
           [getNotAvailableDays [] java.util.List]
           [setNotAvailableDays [java.util.List] void]])

(defn- event-init
  ([] (event-init (java.util.UUID/randomUUID)))
  ([uuid] [[] (ref {:id (java.util.UUID/randomUUID)
                    :external-id uuid
                    :people []
                    :name ""
                    :preferred-slots []
                    :event-day 1
                    :not-available-days []
                    :dependant-event-count 0
                    :event-type EventType/SINGLE})]))

(defn- event-getId [this]
  (getValue this :id))

(defn- event-getExternalId [this]
  (getValue this :external-id))

(defn- event-getSlot [this]
  (getValue this :slot))

(defn- event-setSlot [this item]
  (setValue this :slot item))

(defn- event-getPreferredSlots [this]
  (getValue this :preferred-slots))

(defn- event-setPreferredSlots [this item]
  (setValue this :preferred-slots item))

(defn- event-getChainedEvent [this]
  (getValue this :chained-event))

(defn- event-setChainedEvent [this item]
  (setValue this :chained-event item))

(defn- event-getEventDay [this]
  (getValue this :event-day))

(defn- event-setEventDay [this item]
  (setValue this :event-day item))

(defn- event-getDependantEventCount [this]
  (getValue this :dependant-event-count))

(defn- event-setDependantEventCount [this item]
  (setValue this :dependant-event-count item))

(defn- event-getEventType [this]
  (getValue this :event-type))

(defn- event-setEventType [this item]
  (setValue this :event-type item))

(defn- event-getPeople [this]
  (getValue this :people))

(defn- event-setPeople [this item]
  (setValue this :people item))

(defn- event-getName [this]
  (getValue this :name))

(defn- event-setName [this item]
  (setValue this :name item))

; Technically these are the not-available-days of the *people* attending this event
(defn- event-getNotAvailableDays [this]
  (getValue this :not-available-days))

(defn- event-setNotAvailableDays [this item]
  (setValue this :not-available-days item))

File: src/java/herder/Event.java
package herder.solver.event;

import java.util.UUID;
import java.util.List;
import org.joda.time.Interval;
import org.joda.time.LocalDate;

import org.optaplanner.core.api.domain.variable.CustomShadowVariable;
import org.optaplanner.core.api.domain.entity.PlanningEntity;
import org.optaplanner.core.api.domain.variable.PlanningVariable;

@PlanningEntity(difficultyComparatorClass=EventDifficultyComparator.class)
public class Event implements Cloneable {
    private Event chainedEvent;
    @CustomShadowVariable(variableListenerClass = EventUpdatingVariableListener.class,
            sources = {@CustomShadowVariable.Source(variableName = "slot")})
    public Event getChainedEvent() {return chainedEvent;}
    public void setChainedEvent(Event event) {this.chainedEvent = event;}

    public Event() {
        this(UUID.randomUUID());
    }

    private UUID externalId;
    private UUID id;

    public Event(UUID externalId) {
        this.externalId = externalId;
        this.id = UUID.randomUUID();
    }

    private Interval slot;
    @PlanningVariable(valueRangeProviderRefs={"slotRange"}, nullable=true)
    public Interval getSlot(){return slot;}
    public void setSlot(Interval slot) {this.slot = slot;}

    private LocalDate preferredDay;
    public LocalDate getPreferredDay() {return preferredDay;}
    public void setPreferredDay(LocalDate preferredDay) { this.preferredDay = preferredDay;}

    private List<Interval> preferredSlots;
    public List<Interval> getPreferredSlots(){return preferredSlots;}
    public void setPreferredSlots(List<Interval> preferredSlots) { this.preferredSlots = preferredSlots;}

    private int eventDay;
    public int getEventDay() { return eventDay;}
    public void setEventDay(int eventDay) {this.eventDay = eventDay;}

    private int eventCount;
    public int getDependantEventCount() { return eventCount;}
    public void setDependantEventCount(int eventCount) {this.eventCount = eventCount;}

    private EventType eventType;
    public EventType getEventType() { return eventType;}
    public void setEventType(EventType eventType) {this.eventType = eventType;}

    public UUID getId() { return id;}
    public UUID getExternalId() { return externalId;}

    private List<Object> people;
    public List<Object> getPeople(){return people;}
    public void setPeople(List<Object> people) { this.people = people;}

    private String name;
    public String getName() { return name;}
    public void setName(String name) {this.name = name;}

    private List<LocalDate> notAvailableDays;
    public List<LocalDate> getNotAvailableDays(){return notAvailableDays;}
    public void setNotAvailableDays(List<LocalDate> notAvailableDays) { this.notAvailableDays = notAvailableDays;}

    private Event laterEvent;
    public Event getLaterEvent() {return laterEvent;}
    public void setLaterEvent(Event laterEvent) {this.laterEvent = laterEvent;}

    public Event clone() {
        Event ret = new Event(externalId);
        ret.id = this.id;
        ret.slot = this.slot;
        ret.preferredSlots = this.preferredSlots;
        ret.preferredDay = this.preferredDay;
        ret.eventDay = this.eventDay;
        ret.eventCount = this.eventCount;
        ret.eventType = this.eventType;
        ret.people = this.people;
        ret.name = this.name;
        ret.notAvailableDays = this.notAvailableDays;
        ret.chainedEvent = this.chainedEvent;
        ret.laterEvent = this.laterEvent;
        return ret;
    }

    public void fixChainedEventRef(List<Event> events) {
        if (chainedEvent !=null) {
            for (Event e: events) {
                if (e.id == chainedEvent.id) {
                    chainedEvent = e;
                    break;
                }
            }
        }
        if (laterEvent !=null) {
            for (Event e: events) {
                if (e.id == laterEvent.id) {
                    laterEvent = e;
                }
            }
        }
    }
}
